trie/zkproof/writer.go

```diff
@@ -0,0 +1,731 @@
+				panic("unexpected")
+			}
+
+			//sanity check
+			if accRootFromState := zkt.ReverseByteOrder(statePath[0].Root); !bytes.Equal(acc.Root[:], accRootFromState) {
+				panic(fmt.Errorf("unexpected storage root before: [%s] vs [%x]", acc.Root, accRootFromState))
+			}
+			return &types.StateAccount{
+				Nonce:    acc.Nonce,
+				Balance:  acc.Balance,
+				CodeHash: acc.CodeHash,
+				Root:     common.BytesToHash(zkt.ReverseByteOrder(statePath[1].Root)),
+			}
+		})
+	if err != nil {
+		return nil, fmt.Errorf("update account %s in SSTORE fail: %s", addr, err)
+	}
+
+	if stateUpdate[1] != nil {
+		out.StateKey = statePath[1].Leaf.Sibling
+	} else if stateUpdate[0] != nil {
+		out.StateKey = statePath[0].Leaf.Sibling
+	} else {
+		// it occurs when we are handling SLOAD with non-exist value
+		// still no pretty idea, had to touch the internal behavior in zktrie ....
+		if h, err := storeKey.Hash(); err != nil {
+			return nil, fmt.Errorf("hash storekey fail: %s", err)
+		} else {
+			out.StateKey = zkt.NewHashFromBigInt(h)[:]
+		}
+	}
+
+	out.StatePath = statePath
+	out.StateUpdate = stateUpdate
+	return out, nil
+}
+
+func (w *zktrieProofWriter) HandleNewState(accountState *types.AccountWrapper) (*StorageTrace, error) {
+
+	if accountState.Storage != nil {
+		storeAddr := hexutil.MustDecode(accountState.Storage.Key)
+		storeValue := hexutil.MustDecode(accountState.Storage.Value)
+		return w.traceStorageUpdate(accountState.Address, storeAddr, storeValue)
+	} else {
+
+		accData := getAccountDataFromLogState(accountState)
+
+		out, err := w.traceAccountUpdate(accountState.Address, func(accBefore *types.StateAccount) *types.StateAccount {
+			if accBefore != nil {
+				accData.Root = accBefore.Root
+			}
+			return accData
+		})
+		if err != nil {
+			return nil, fmt.Errorf("update account state %s fail: %s", accountState.Address, err)
+		}
+		hash, err := zkt.NewHashFromBytes(accData.Root[:])
+		if err != nil {
+			return nil, fmt.Errorf("malform of state root in account %s", accountState.Address)
+		}
+		out.CommonStateRoot = hash[:]
+		return out, nil
+	}
+
+}
+
+func handleLogs(od opOrderer, currentContract common.Address, logs []*types.StructLogRes) {
+	logStack := []int{0}
+	contractStack := map[int]common.Address{}
+	skipDepth := 0
+	callEnterAddress := currentContract
+
+	// now trace every OP which could cause changes on state:
+	for i, sLog := range logs {
+
+		//trace log stack by depth rather than scanning specified op
+		if sl := len(logStack); sl < sLog.Depth {
+			logStack = append(logStack, i)
+			//update currentContract according to previous op
+			contractStack[sl] = currentContract
+			currentContract = callEnterAddress
+		} else if sl > sLog.Depth {
+			logStack = logStack[:sl-1]
+			currentContract = contractStack[sLog.Depth]
+			resumePos := logStack[len(logStack)-1]
+			calledLog := logs[resumePos]
+
+			//no need to handle fail calling
+			if calledLog.ExtraData == nil || !calledLog.ExtraData.CallFailed {
```

Comment:
I think
```suggestion
			if !(calledLog.ExtraData != nil && calledLog.ExtraData.CallFailed) {
```

is more readable